services/report/__init__.py

```diff
@@ -1239,3 +1255,31 @@ def save_parallel_report_to_archive(
             "chunks_path": chunks_url,
             "files_and_sessions_path": files_and_sessions_url,
         }
+
+
+@sentry_sdk.trace
+def delete_uploads_by_sessionid(upload: Upload, session_ids: list[int]):
```

Comment:
3 things here: 

1) I could see this taking some time because some customers have a gargantuan number of CFFs in their commits, so this might benefit from making a standalone task - unless we're wanting to do this synchronously. In which case, we'd take a bit extra time processing, so something to contemplate/measure in some way if we can
2) How could this react with other sessions being processed in parallel? Not sure if while deleting we could run into errors if two or more processes are trying to dleete the same entry
3) How could we test this in practice? I just want to make sure we aren't silently corrupting data by not acknowledging our reports were using data from cff's we deleted or something like that

Reply:
1. not quite sure the overhead of a separate task (both complexity and runtime) is reasonable for that. but you are right that this might be quite slow. However it is limited to only those uploads/sessions which have been overridden by an upload, so that should be rather limited.
2. we are behind the "modify report" (upload processing) lock, so there is no parallelism here. also concurrent deletes shouldn’t be a problem either way, as we filter things by `IN`, which sql just ignores when no matching rows exist.
3. good point, I don’t have a good idea for this yet. Though this change aims to align data in SQL with data inside of the `report_json`. So this is actually fixing a pre-existing "data corruption" if you think about it that way.

with these questions, you have brought up another good point: do we need to care about SQL-level locking here? I don’t think we need to, as the deletes are only locking the effected rows for the duration of the transaction (which I guess spans the whole task?), and those particular rows are only touched behind the upload processing lock.

Reply:
Assuming the deletions take place a short amount of time, I don't think we'd need to worry